File: game.lua
local Object = require "lib.classic"
local decks = require "decks"
local cards = require "cards"

---@class Game
---@field private decks Deck[]
---@field private flatDecks FlatDeck[]
---@field private homes HomeDeck[]
---@field private reserve ReserveDeck
---@field draw fun(self:Game)
---@field new_game fun(self:Game)
---@field debug number
---@overload fun():Game
local Game = Object:extend()

function Game:new()
    self.flatDecks = {
        decks.FlatDeck(2 + 50 * 0, 70),
        decks.FlatDeck(2 + 50 * 1, 70),
        decks.FlatDeck(2 + 50 * 2, 70),
        decks.FlatDeck(2 + 50 * 3, 70),
        decks.FlatDeck(2 + 50 * 4, 70),
        decks.FlatDeck(2 + 50 * 5, 70),
        decks.FlatDeck(2 + 50 * 6, 70)
    }
    self.homes = {
        decks.HomeDeck(152 + 50 * 0, 2, cards.Suit.Hearts),
        decks.HomeDeck(152 + 50 * 1, 2, cards.Suit.Diamonds),
        decks.HomeDeck(152 + 50 * 2, 2, cards.Suit.Clubs),
        decks.HomeDeck(152 + 50 * 3, 2, cards.Suit.Spades)
    }
    self.reserve = decks.ReserveDeck(2, 2)
    self.decks = { self.reserve }
    for _, flat in ipairs(self.flatDecks) do
        table.insert(self.decks, flat)
    end
    for _, home in ipairs(self.homes) do
        table.insert(self.decks, home)
    end
    self.debug = 0
end

function Game:draw()
    for _, deck in ipairs(self.decks) do
        deck:draw()
    end

    if self.debug == 1 then
        love.graphics.setColor(1, 0.2, 0.2)
        for _, deck in ipairs(self.decks) do
            local bounds = deck.boundsGrab
            if bounds then
                love.graphics.rectangle("line", bounds.x, bounds.y, bounds.w, bounds.h)
            end
        end
        love.graphics.setColor(1, 1, 1)
    end
end

---@param card_list Card[]
local function shuffle(card_list)
    for i = 1, #card_list - 1 do
        local r = love.math.random(i, #card_list)
        card_list[i], card_list[r] = card_list[r], card_list[i]
    end
end

function Game:new_game()
    for _, deck in ipairs(self.decks) do
        deck:clear()
    end

    local main_deck = {}
    for suit = 1, 4 do
        for rank = 1, 13 do
            table.insert(main_deck, cards.Card(suit, rank))
        end
    end
    shuffle(main_deck)

    for i, base_deck in ipairs(self.flatDecks) do
        for j = 1, i do
            table.insert(base_deck.cards, table.remove(main_deck))
        end
        base_deck.covered = #base_deck.cards - 1
    end

    self.reserve.cards = main_deck
end

return Game

File: src/game.ts
import { Card, Suit } from "./cards";
import { Deck, FlatDeck, HomeDeck, ReserveDeck } from "./decks";
import { Vec2 } from "./geometry";

function shuffle(cardList: Card[]) {
    for (const i of $range(0, cardList.length - 2)) {
        const r = love.math.random(i, cardList.length - 1);
        cardList[i], cardList[r] = cardList[r], cardList[i];
    }
}

export class Game {
    decks: Deck[];
    flatDecks: FlatDeck[];
    homes: HomeDeck[];
    reserve: ReserveDeck;
    mousePos: Vec2;
    debugBounds: number;

    constructor() {
        this.flatDecks = [
            new FlatDeck(2 + 50 * 0, 70),
            new FlatDeck(2 + 50 * 1, 70),
            new FlatDeck(2 + 50 * 2, 70),
            new FlatDeck(2 + 50 * 3, 70),
            new FlatDeck(2 + 50 * 4, 70),
            new FlatDeck(2 + 50 * 5, 70),
            new FlatDeck(2 + 50 * 6, 70)
        ]
        this.homes = [
            new HomeDeck(152 + 50 * 0, 2, Suit.Hearts),
            new HomeDeck(152 + 50 * 1, 2, Suit.Diamonds),
            new HomeDeck(152 + 50 * 2, 2, Suit.Clubs),
            new HomeDeck(152 + 50 * 3, 2, Suit.Spades)
        ]
        this.reserve = new ReserveDeck(2, 2);
        this.decks = [this.reserve, ...this.flatDecks, ...this.homes];
        this.debugBounds = 0;
        this.mousePos = new Vec2(0, 0);
    }

    draw() {
        for (const deck of this.decks) {
            deck.draw();
        }

        switch (this.debugBounds) {
            case 1:
                love.graphics.setColor(1, 0.2, 0.2);
                for (const deck of this.decks) {
                    const bounds = deck.boundsGrab;
                    if (bounds) love.graphics.rectangle("line", bounds.x, bounds.y, bounds.w, bounds.h);
                }
                love.graphics.setColor(1, 1, 1);
                break;

            case 2:
                love.graphics.setColor(1, 0.2, 0.2);
                for (const deck of this.decks) {
                    const bounds = deck.boundsDrop;
                    if (bounds) love.graphics.rectangle("line", bounds.x, bounds.y, bounds.w, bounds.h);
                }
                love.graphics.setColor(1, 1, 1);
                break;

            case 3:
                love.graphics.setColor(1, 0.2, 0.2);
                for (const deck of this.decks) {
                    const bounds = deck.boundsClick;
                    if (bounds) love.graphics.rectangle("line", bounds.x, bounds.y, bounds.w, bounds.h);
                }
                love.graphics.setColor(1, 1, 1);
                break;

            case 4:
                love.graphics.setColor(1, 0.2, 0.2);
                for (const deck of this.decks) {
                    const bounds = deck.boundsDblClick;
                    if (bounds) love.graphics.rectangle("line", bounds.x, bounds.y, bounds.w, bounds.h);
                }
                love.graphics.setColor(1, 1, 1);
                break;
        }
    }

    newGame() {
        for (const deck of this.homes) {
            deck.clear();
        }

        const mainDeck: Card[] = [];
        for (const suit of $range(0, 3)) {
            for (const rank of $range(0, 12)) {
                mainDeck.push(new Card(suit, rank));
            }
        }
        shuffle(mainDeck);

        this.flatDecks.forEach((deck, i) => {
            deck.cards = mainDeck.splice(0, i + 1);
            deck.covered = deck.cards.length - 2;
        });

        this.reserve.cards = mainDeck;

        for (const deck of this.decks) {
            deck.updateBounds();
        }
    }
}